designs/full-cluster-lifecycle-api-cloudstack.md

```diff
@@ -0,0 +1,81 @@
+* We only support Redhat, not on Ubuntu or Bottlerocket yet.
+
+## Overview of Solution
+
+The EKS-A controller running in management cluster reconciles EKS-A CRDs, which fully manages workload clusters. [Detailed lifecycle diagram](https://github.com/aws/eks-anywhere/blob/4d307c6fe75075adadae38537d0f211c9142003e/designs/full-cluster-lifecycle-api.md#L55)
+
+### CloudStack defaults
+
+We already have [SetDefaults](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/api/v1alpha1/cloudstackdatacenterconfig_types.go#L216) for datacenter, we can utilize it in its mutation webhook (need to create one). Cloudstack machine config doesn’t have machine config [default](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/providers/cloudstack/cloudstack.go#L162) values at this stage.
+
+### Static/Data validation
+
+CloudStack already has webhook for datacenter and machine config. We’ll reorganize validations, so keep static/data validations in webhook and keep validations need cmk in validator. The validator will be used in CloudStack provider, cluster reconciler and datacenter reconciler.
+
+* Extract static/data validation from [validateDatacenterConfig](https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/validator.go#L60) and ValidateClusterMachineConfigs (https://github.com/aws/eks-anywhere/blob/main/pkg/providers/cloudstack/validator.go#L127) to datacenterConfig/machineConfig types, and called in webhook. The runtime validations stay in validator and will be called during reconcile.
+* We have immutable fields validation in both webhook ([validateImmutableFieldsCloudStackMachineConfig](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go#L86) and [validateImmutableFieldsCloudStackCluster](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go#L109)) and provider ([ValidateClusterMachineConfigs](https://github.com/aws/eks-anywhere/blob/3c1fd0ff732641ed02137213863942403f59c320/pkg/providers/cloudstack/validator.go#L114)). There’re some gaps between two places so we’ll need to decide the final validation and cleanup duplicate/mismatch logic.
+
+### Runtime Validation
+
+#### Validator Initialization
+
+* **How to read credential**
+
+  The validator requires cloudstack credential to build cmk. Our current CLI reads credential from env, and has a validation that accept new credential but not allowing existing credential update.
+
+    1. Read credentials from existing secrets which requires customers to add additional secrets. We also need to add a mechanism to prevent customers from update existing secrets.
+    2. (preferred) Read credentials from env like current CLI flow, which could prevent customers from updating credentials and also is an easy way to add new credentials.
```

Comment:
This is a bad idea
In fact, it won't work

Different clusters can have different sets of credentials, and if you inject the credentials in the controller through env vars, you can only have one fixed set. Moreover, this makes impossible for the user to rotate credentials.

I would advise to look into the general direction we recently gave for all providers on how to implement credentials. And try to follow it if possible.

Reply:
sry, I meant reading from credential file. Using secrets would make more sense as we don't need to prevent users from updating secrets.